sys/drv/umass.c

```diff
@@ -0,0 +1,464 @@
+ */
+
+static inline usb_direction_t uioop_to_usbdir(uio_op_t op) {
+  switch (op) {
+    case UIO_READ:
+      return USB_DIR_INPUT;
+    case UIO_WRITE:
+      return USB_DIR_OUTPUT;
+    default:
+      panic("Invalid UIO operation!");
+  }
+}
+
+static int umass_op(vnode_t *v, uio_t *uio) {
+  device_t *dev = devfs_node_data(v);
+  umass_state_t *umass = dev->state;
+  usb_direction_t dir = uioop_to_usbdir(uio->uio_op);
+  uint32_t block_size = umass->block_size;
+  int error = 0;
+
+  if (!is_aligned(uio->uio_offset, block_size))
+    return EINVAL;
+  if (!is_aligned(uio->uio_resid, block_size))
+    return EINVAL;
+
+  uint32_t start = uio->uio_offset / block_size;
+  uint32_t nblocks = uio->uio_resid / block_size;
+  uint32_t size = nblocks * block_size;
+
+  /* Note: we assume that number of blocks to read is <= UINT16_MAX. */
+  if (nblocks > UINT16_MAX)
+    return EINVAL;
+
+  void *buf = kmalloc(M_DEV, size, M_WAITOK);
+
+  if (dir == USB_DIR_OUTPUT) {
+    /* Copy data from the user. */
+    if ((error = uiomove(buf, uio->uio_resid, uio)))
+      goto end;
+  }
+
+  scsi_rw_10_t rw10 = (scsi_rw_10_t){
+    .opcode = (dir == USB_DIR_INPUT) ? READ_10 : WRITE_10,
+    .addr = htobe32(start),
+    .length = htobe16(nblocks),
+  };
+  error = umass_transfer(dev, &rw10, sizeof(scsi_rw_10_t), dir, buf, size);
+  if (error)
+    goto end;
+
+  uio->uio_offset = 0;
+
+  if (dir == USB_DIR_INPUT) {
+    /* Copy the data to the user. */
+    error = uiomove_frombuf(buf, size, uio);
+  }
+
+end:
+  kfree(M_DEV, buf);
+  return error;
+}
+
+static vnodeops_t umass_ops = {
+  .v_read = umass_op,
+  .v_write = umass_op,
+};
+
+/*
+ * Driver interface implementation.
+ */
+
+static int umass_probe(device_t *dev) {
+  usb_device_t *udev = usb_device_of(dev);
+  if (!udev)
+    return 0;
+
+  /* We're looking for a USB mass storage device using SCSI transparent
+   * command set and Bulk-Only transfer protocol. */
+  if (udev->class_code != UICLASS_MASS ||
+      udev->subclass_code != UISUBCLASS_SCSI ||
+      udev->protocol_code != UIPROTO_MASS_BBB)
+    return 0;
+
+  return 1;
+}
+
+static int umass_attach(device_t *dev) {
+  int error = 0;
+
+  /* Obtain the max Logical Unit Number of the drive. */
+  uint8_t maxlun;
+  if (usb_bbb_get_max_lun(dev, &maxlun))
+    return ENXIO;
+
+  /* We only support drives with a single logical unit. */
+  if (maxlun)
+    return ENXIO;
+
+  /* Set the initial block size. */
+  umass_state_t *umass = dev->state;
+  umass->block_size = UMASS_MIN_BLOCK_SIZE;
+
+  /* Identify the connected device. */
+  if ((error = umass_inquiry(dev)))
+    return ENXIO;
+
+  /*
+   * After receiving a successful inquiry transfer,
+   * the drive is ready for use.
+   */
+
+  /* Obtain the number and length of the logical blocks. */
+  if ((error = umass_read_capacity(dev)))
+    return ENXIO;
+
+  umass_print(dev);
+
+  /* Prepare /dev/umass interface. */
+  devfs_makedev(NULL, "umass", &umass_ops, dev, NULL);
```

Comment:
We're trying to get rid of this function. Please replace it with `devfs_makedev_new`.

Reply:
Done.